src/Control/Monad/Eff/AVar.js

```diff
@@ -0,0 +1,318 @@
+  this.takes    = new MutableQueue();
+  this.reads    = new MutableQueue();
+  this.puts     = new MutableQueue();
+}
+
+exports.makeEmptyVar = function () {
+  return new AVar();
+};
+
+exports.makeVar = function (value) {
+  return function () {
+    var avar = new AVar();
+    avar.value = value;
+    return avar;
+  };
+};
+
+exports._killVar = function (left, right, avar, error) {
+  return function () {
+    if (avar.error === null) {
+      avar.error = error;
+      avar.value = EMPTY;
+      drainVar(left, right, avar);
+    }
+  };
+};
+
+exports._putVar = function (left, right, avar, value, cb) {
+  return function () {
+    if (avar.error !== null) {
+      runEff(cb(left(avar.error)));
+      return NO_EFFECT;
+    }
+    var cell = putLast(avar.puts, { cb: cb, value: value });
+    drainVar(left, right, avar);
+    return function () {
+      deleteCell(cell);
+    };
+  };
+};
+
+exports._takeVar = function (left, right, avar, cb) {
+  return function () {
+    if (avar.error !== null) {
+      runEff(cb(left(avar.error)));
+      return NO_EFFECT;
+    }
+    var cell = putLast(avar.takes, cb);
+    drainVar(left, right, avar);
+    return function () {
+      deleteCell(cell);
+    };
+  };
+};
+
+exports._readVar = function (left, right, avar, cb) {
+  return function () {
+    if (avar.error !== null) {
+      runEff(cb(left(avar.error)));
+      return NO_EFFECT;
+    }
+    var cell = putLast(avar.reads, cb);
+    drainVar(left, right, avar);
+    return function () {
+      deleteCell(cell);
+    };
+  };
+};
+
+exports._tryPutVar = function (left, right, avar, value) {
+  return function () {
+    if (avar.value === EMPTY && avar.error === null) {
+      avar.value = value;
+      drainVar(left, right, avar);
+      return true;
+    } else {
+      return false;
+    }
+  };
+};
+
+exports._tryTakeVar = function (left, right, nothing, just, avar) {
+  return function () {
+    var value = avar.value;
+    if (value === EMPTY) {
+      return nothing;
+    } else {
+      avar.value = EMPTY;
+      drainVar(left, right, avar);
+      return just(value);
+    }
+  };
+};
+
+exports._tryReadVar = function (nothing, just, avar) {
+  return function () {
+    if (avar.value === EMPTY) {
+      return nothing;
+    } else {
+      return just(avar.value);
+    }
+  };
+};
+
+exports.isEmptyVar = function (avar) {
+  return function () {
+    return avar.value === EMPTY;
+  };
+};
+
+function drainVar (left, right, avar) {
+  if (avar.draining) {
+    return;
+  }
+
+  var ps = avar.puts;
+  var ts = avar.takes;
+  var rs = avar.reads;
+  var p, r, t, value, rsize;
+
+  avar.draining = true;
+
+  /* jshint -W084 */
+  while (1) {
+    p = null;
+    r = null;
+    t = null;
+    value = avar.value;
+    rsize = rs.size;
+
+    if (avar.error !== null) {
+      value = left(avar.error);
+      // Error callback ordering is somewhat undefined, but we try to at least
+      // be fair by interleaving puts and takes.
```

Comment:
If the logic of take/put is correct, then this is overgeneralized and supports more possibilities than can actually exist.

```haskell
data AVarState a 
  = Empty 
  | Full a 
  | Surplus (NonEmptyList (Tuple a (Callback Unit))) 
  | Deficit (NonEmptyList (Callback a)) (List (Callback a)) -- takers, readers
```

Taking advantage of these constraints could simplify things and remove the need to handle "impossible cases", as well as provide stronger guarantees. 

Note that `NonEmptyList` above should actually be something like `NonEmptyFifoQueue`.